contributors/design-proposals/containerized-mounter-pod.md

```diff
@@ -0,0 +1,145 @@
+
+To sum it up, it's just a daemon set that spawns privileged pods with some labels, running a simple init and waiting for Kubernetes to do `kubectl exec <the pod> <some utility> <args>`.
+
+## Design
+
+### Configuration of the host OS
+With the [HostPath volume propagation](propagation.md) implemented, we must ensure that `/var/lib/kubelet` is share-able into containers so a pod with mount utilities can mount something there in its container and this mount will be visible to kubelet and docker (or other container engine) on the host.
+We propose:
+*  During startup. kubelet checks that `/var/lib/kubelet` on the host is on a mount with shared mount propagation.
+    *  If not, kubelet makes `/var/lib/kubelet` shareable:
+       ```shell
+       mount --bind /var/lib/kubelet /var/lib/kubelet
+       mount --make-shared /var/lib/kubelet
+       ```
+       Note that these commands are executed on the host in case kubelet runs in a container.
+       If this startup adjustment fails, kubelet refuses to start with a clear error message.
+* Kubelet does not do any other special check. We assume that kubelet either runs on the host and in the same mount namespace as docker and the host, or it runs in a container that sees /var/lib/kubelet as it is on the host and with shared mount propagation.
+
+This ensures that kubelet runs out of the box on any distro without any configuration done by the cluster admin.
+
+### Volume plugins
+* All volume plugins need to be updated to use a new `VolumeExec` interface to call external utilities like `mount`, `mkfs`, `rbd lock` and such. Implementation of the interface will be provided by caller and will lead either to `exec` on the host or `kubectl exec` or `docker exec` in a remote or local pod with utilities for appropriate volume plugin (or docker-exec-like command if another container engine is used).
+
+### Controller
+* There will be new parameter to kube-controller-manager and kubelet:
+    * `--experimental-mount-namespace`, which specifies a dedicated namespace where all pods with mount utilities reside. It would default to `kube-mount`.
+* Whenever PV or attach/detach controller needs to call a volume plugin, it looks for *any* running pod in the specified namespace with label `mount.kubernetes.io/foo=true` (or `mount.kubernetes.io/flexvolume/foo=true` for flex volumes) and calls the volume plugin so it all mount utilities are executed as `kubectl exec <pod> xxx` (of course, we'll use clientset interface instead of executing `kubectl`).
+* If such pod does not exist, it executes the mount utilities on the host as usual.
+* During alpha, no controller-manager changes will be done. That means that Ceph RBD provisioner will still require `/usr/bin/rbd` installed on the master. All other volume plugins will work without any problem, as they don't execute any utility when attaching/detaching/provisioning/deleting a volume.
```

Comment:
the rbd provisioner has been pulled out to here:
https://github.com/kubernetes-incubator/external-storage/tree/master/ceph/rbd

so the container can be built with the right ceph version already.